libs/postgres_ffi/wal_generate/src/lib.rs

```diff
@@ -0,0 +1,287 @@
+        Ok(self
+            .query_one("SELECT pg_current_wal_insert_lsn()", &[])?
+            .get(0))
+    }
+    fn pg_current_wal_flush_lsn(&mut self) -> Result<PgLsn> {
+        Ok(self
+            .query_one("SELECT pg_current_wal_flush_lsn()", &[])?
+            .get(0))
+    }
+}
+
+impl<C: postgres::GenericClient> PostgresClientExt for C {}
+
+pub fn generate_simple(client: &mut impl postgres::GenericClient) -> Result<PgLsn> {
+    let initial_lsn = client.pg_current_wal_insert_lsn()?;
+    info!("LSN initial = {}", initial_lsn);
+    client.execute("CREATE table t(x int)", &[])?;
+    let last_lsn = client.pg_current_wal_insert_lsn()?;
+    ensure!(
+        last_lsn == client.pg_current_wal_flush_lsn()?,
+        "WAL is either not flushed or is extended with something unknown"
+    );
+    Ok(last_lsn)
+}
+
+pub fn generate_wal_record_crossing_segment_followed_by_small_one(
+    client: &mut impl postgres::GenericClient,
+) -> Result<PgLsn> {
+    let initial_lsn = client.pg_current_wal_insert_lsn()?;
+    info!("LSN initial = {}", initial_lsn);
+    ensure!(
+        initial_lsn < PgLsn::from(0x0200_0000 - 4 * 8192),
+        "Initial LSN is too far in the future"
+    );
+
+    // This message will be followed by a small COMMIT message and flushed for sure.
+    // Non-transactional messages may be not flushed immediately and wait for the background thread (which flushes every `wal_writer_delay` ms) or following WAL.
+    let message_lsn: PgLsn = client.query_one(
+        "select pg_logical_emit_message(true, 'big-17mb-msg', concat(repeat('abcd', 17 * 256 * 1024), 'end')) as message_lsn",
+        &[]
+    )?.get("message_lsn");
+    ensure!(
+        message_lsn > PgLsn::from(0x0200_0000 + 4 * 8192),
+        "Logical message did not cross the segment boundary"
+    );
+
+    let after_message_lsn = client.pg_current_wal_insert_lsn()?;
+    ensure!(
+        message_lsn < after_message_lsn,
+        "No record found after the emitted message"
+    );
+
+    ensure!(
+        after_message_lsn == client.pg_current_wal_flush_lsn()?,
+        "WAL is either not flushed or is extended with something unknown"
+    );
+    Ok(after_message_lsn)
+}
+
+pub fn generate_last_wal_record_crossing_segment<C: postgres::GenericClient>(
+    client: &mut C,
+) -> Result<PgLsn> {
+    // First few created tables take more WAL bytes than later.
+    info!("LSN initial = {}", client.pg_current_wal_insert_lsn()?);
+    client.execute("create table t_base_1 (x int)", &[])?;
+    client.execute("create table t_base_2 (x int)", &[])?;
+
+    // We want to create ~400 tables: ~4300 bytes in WAL per each, ~353 bytes in the COMMIT message per each.
+    // Commit message will take approx 137 KiB and will start ~24 KiB before segment's end.
+    let start_creating_at_lsn = PgLsn::from(0x0200_0000 - 8192 * 3 - 4300 * 400);
+    let stop_creating_at_lsn = PgLsn::from(0x0200_0000 - 8192 * 3);
+    let expect_commit_end_after_lsn = PgLsn::from(0x0200_0000);
```

Comment:
Why it is so complicated, what is the problem with single large record crossing the segment?

Reply:
The original problem was that there was no way to ensure WAL is flushed to the disk. No longer a problem with `XLogFlush` exposed as a SQL function